numpy/_array_api/_array_object.py

```diff
@@ -0,0 +1,891 @@
+                raise TypeError("Python bool scalars can only be promoted with bool arrays")
+        elif isinstance(scalar, int):
+            if self.dtype in _boolean_dtypes:
+                raise TypeError("Python int scalars cannot be promoted with bool arrays")
+        elif isinstance(scalar, float):
+            if self.dtype not in _floating_dtypes:
+                raise TypeError("Python float scalars can only be promoted with floating-point arrays.")
+        else:
+            raise TypeError("'scalar' must be a Python scalar")
+
+        # Note: the spec only specifies integer-dtype/int promotion
+        # behavior for integers within the bounds of the integer dtype.
+        # Outside of those bounds we use the default NumPy behavior (either
+        # cast or raise OverflowError).
+        return ndarray._new(np.array(scalar, self.dtype))
+
+    @staticmethod
+    def _normalize_two_args(x1, x2):
+        """
+        Normalize inputs to two arg functions to fix type promotion rules
+
+        NumPy deviates from the spec type promotion rules in cases where one
+        argument is 0-dimensional and the other is not. For example:
+
+        >>> import numpy as np
+        >>> a = np.array([1.0], dtype=np.float32)
+        >>> b = np.array(1.0, dtype=np.float64)
+        >>> np.add(a, b) # The spec says this should be float64
+        array([2.], dtype=float32)
+
+        To fix this, we add a dimension to the 0-dimension array before passing it
+        through. This works because a dimension would be added anyway from
+        broadcasting, so the resulting shape is the same, but this prevents NumPy
+        from not promoting the dtype.
+        """
+        if x1.shape == () and x2.shape != ():
+            # The _array[None] workaround was chosen because it is relatively
+            # performant. broadcast_to(x1._array, x2.shape) is much slower. We
+            # could also manually type promote x2, but that is more complicated
+            # and about the same performance as this.
+            x1 = ndarray._new(x1._array[None])
+        elif x2.shape == () and x1.shape != ():
+            x2 = ndarray._new(x2._array[None])
+        return (x1, x2)
+
+    # Everything below this line is required by the spec.
+
+    def __abs__(self: array, /) -> array:
+        """
+        Performs the operation __abs__.
+        """
+        res = self._array.__abs__()
+        return self.__class__._new(res)
+
+    @np.errstate(all='ignore')
+    def __add__(self: array, other: Union[int, float, array], /) -> array:
+        """
+        Performs the operation __add__.
+        """
+        if isinstance(other, (int, float, bool)):
+            other = self._promote_scalar(other)
+        self, other = self._normalize_two_args(self, other)
+        res = self._array.__add__(other._array)
+        return self.__class__._new(res)
+
+    def __and__(self: array, other: Union[int, bool, array], /) -> array:
```

Comment:
A question here: does NEP 47 currently limit itself to builtin scalar types, or are the likes of `np.generic` also allowed?

Actually, come to think of it, `np.generic` seems to fall under the `array` umbrella here, looking at the source code.

Reply:
This is on purpose, `np.generic` and custom scalars are not included, with the main reason being that those numerical scalars are not portable. Also that would include things like `datetime64`, `object`, etc. that are definitely out of scope.

There's also a more principled design question here. The NumPy type annotations are correct for NumPy, but they are also pretty much incomprehensible. Which is a sign that something went wrong in the design. Array scalars are considered a design mistake, and you also wouldn't build this huge pyramid of abstract dtype classes.

I think from the array API standard point of view, things should be typed clearly and unambiguously - which in this case means only Python scalars. If, as an implementer, you have a good reason to implement a custom object that is `float`-like or `tuple`-like, then you can still do so within your own library, you just have to make sure it duck types correctly both at runtime and at type-check time (can be done by making your custom object a subtype).